### PR TITLE
Initialize tfexec.Terraform once per request + refactoring/cleanup

### LINKED ISSUE #6456: Consolidate multiple initializations of tfexec.Terraform + logging configuration
### Observed behavior (required)
During Terraform recipe deployment we are [initializing tfexec.Terraform instance and configuring logging](https://github.com/radius-project/radius/blob/main/pkg/recipes/terraform/types.go#L59) twice. Once during [download module](https://github.com/radius-project/radius/blob/main/pkg/recipes/terraform/execute.go#L307) and second during [tf init](https://github.com/radius-project/radius/blob/main/pkg/recipes/terraform/execute.go#L355).

### Proposed Fix (optional)
We should initialize and configure logging once and pass it to the functions that consume it.

[AB#9699](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/9699)

### PR DESCRIPTION
**Pull request by kachawla**
# Description

* Terraform instance along with logging configuration is being called multiple times for same request - consolidating it to be initialized only once.

* Move working directory creation to tfexec.Terraform initialization to remove duplicated code.

## Type of change

- This pull request fixes a bug in Radius and has an approved issue (issue link required).

Fixes: https://github.com/radius-project/radius/issues/6456

### Summary
📝🛠️🚀

Refactored the terraform package to use the `tfexec.Terraform` type and a custom logger. Improved the consistency and readability of the tests and the code. Added new files `log.go` and `types_test.go` to implement and test the new features.

> _Oh we're the `terraform` crew and we've got work to do_
> _We'll refactor and improve our code with `tfexec` too_
> _We'll stream our logs and test our types and make our imports clear_
> _So heave away, me hearties, heave away with cheer_

### Walkthrough
*  Fix import alias for datastoresrp package in `show_test.go` ([link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-13fc0c59e89a1fa92926d4fdd0ef9a6ef372c9816c6c43ec6b003427b63e0852L33-R33), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-13fc0c59e89a1fa92926d4fdd0ef9a6ef372c9816c6c43ec6b003427b63e0852L49-R49), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-13fc0c59e89a1fa92926d4fdd0ef9a6ef372c9816c6c43ec6b003427b63e0852L58-R58), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-13fc0c59e89a1fa92926d4fdd0ef9a6ef372c9816c6c43ec6b003427b63e0852L67-R67), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-13fc0c59e89a1fa92926d4fdd0ef9a6ef372c9816c6c43ec6b003427b63e0852L76-R76), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-13fc0c59e89a1fa92926d4fdd0ef9a6ef372c9816c6c43ec6b003427b63e0852L114-R114), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-13fc0c59e89a1fa92926d4fdd0ef9a6ef372c9816c6c43ec6b003427b63e0852L148-R148), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-13fc0c59e89a1fa92926d4fdd0ef9a6ef372c9816c6c43ec6b003427b63e0852L190-R190), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-13fc0c59e89a1fa92926d4fdd0ef9a6ef372c9816c6c43ec6b003427b63e0852L225-R225))
*  Refactor Terraform executor creation and configuration in `types.go` and `log.go` ([link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-0ad85ed086350c77c3e366408120b4c3020d1cec542671ab3586fd0c34dafe6fL22-R25), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-0ad85ed086350c77c3e366408120b4c3020d1cec542671ab3586fd0c34dafe6fL30-R37), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-0ad85ed086350c77c3e366408120b4c3020d1cec542671ab3586fd0c34dafe6fL59-R71), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-0ad85ed086350c77c3e366408120b4c3020d1cec542671ab3586fd0c34dafe6fL71-R92), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-cd97896943008b717cf4f1a771f6dd35263008aebb0ebccb272755793d10a1acR1-R87), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-8242ca1f206aeefb27fdc2d73686357bf7cb70572c1003c0586e2ba189a6e7f6R1-R56))
*  Simplify generateConfig, initAndApply, initAndDestroy, downloadAndInspect, and downloadModule functions to accept a Terraform executor instead of a working directory and an exec path in `execute.go` and `module.go` ([link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L23-R26), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L45-L49), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L92-R86), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L99-R92), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L105-R98), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L141-R135), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L148-R141), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L168-R161), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L201-R205), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L222-R218), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L243-R225), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L301-R283), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L307-R289), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L320-R302), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L353-R339), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-41ec31c0ceb8bd50c0329b55a8d27aeba69648059faef4e785302155f4e97482L385-R371), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-4f4cb2009f77e8f5e252eee95f3ed880148fc285798ef4b7bbeb970525ad7b5aR25), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-4f4cb2009f77e8f5e252eee95f3ed880148fc285798ef4b7bbeb970525ad7b5aL101-R103))
*  Update TestGeneratedConfig, Test_GetTerraformConfig, Test_GetTerraformConfig_EmptyRecipeName, and Test_GetTerraformConfig_InvalidDirectory functions to use the NewTerraform function and pass the test directory instead of the working directory in `execute_test.go` ([link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL20-R23), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL30-R30), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL78-R38), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL85-L105), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL115-R58), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL128-L129), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL142-R81), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL151-L152), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL161-R98), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL167-R104), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL179-R115))
*  Remove Test_DownloadModule_EmptyWorkingDirPath_Error test case as it is no longer relevant in `module_test.go` ([link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-35f143c273df8b35733ba5b5e44bc5bb5e4dbe8467aeabc3e861b5e19354791bL20-R22), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-35f143c273df8b35733ba5b5e44bc5bb5e4dbe8467aeabc3e861b5e19354791bL92-L101))
*  Rename test case for empty recipe name error in `execute_test.go` ([link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL78-R38))
*  Change test directory name and error message in Test_GetTerraformConfig_InvalidDirectory function in `execute_test.go` ([link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL167-R104), [link](https://github.com/radius-project/radius/pull/6630/files?diff=unified&w=0#diff-3b5890b2cb244038dacda842a1e285235330c421ad27e8bc02d6f979ee8869caL179-R115))

### CODE REVIEW

File: pkg/recipes/terraform/execute.go [hunk 1]
```diff
@@ -89,20 +82,20 @@ func (e *executor) Deploy(ctx context.Context, options Options) (*tfjson.State,
 		return nil, err
 	}
 
-	// Create Working Directory
-	workingDir, err := createWorkingDir(ctx, options.RootDir)
+	// Create a new instance of tfexec.Terraform with current Terraform installation path
+	tf, err := NewTerraform(ctx, options.RootDir, execPath)
```
Comment (ytimocin) [inline pkg/recipes/terraform/execute.go:86]: Could working in the Terraform installation path cause clashes with other files that will take place in the same directory?

Before we were creating a working directory (**workingDir**) in the root directory, now we will not create a working directory but will work in the execution path (**execPath**) -- the working directory and execution path will be the same. Would that cause any problems?

File: pkg/recipes/terraform/module.go [hunk 2]
```diff
@@ -98,13 +99,8 @@ func inspectModule(workingDir, localModuleName string) (*moduleInspectResult, er
...
 // It uses Terraform's Get command to download the module using the Terraform executable available at execPath.
 // An error is returned if the module could not be downloaded.
-func downloadModule(ctx context.Context, workingDir, execPath, templatePath string) error {
-	tf, err := NewTerraform(ctx, workingDir, execPath)
-	if err != nil {
-		return err
-	}
-
-	if err = tf.Get(ctx); err != nil {
+func downloadModule(ctx context.Context, tf *tfexec.Terraform, templatePath string) error {
```
Comment (ytimocin) [inline pkg/recipes/terraform/module.go:102]: I think that, in the future, we could have a struct that has **tfexec.Terraform** as a property. That could have a constructor and some functions. One of the functions could be **downloadModule**, which could be something like this:

```
func (td *TerraformDriver) downloadModule(ctx context.Context, templatePath string) error {
         td.Terraform.Get(ctx)...
...
}
```

Note: The struct name will probably something other than TerraformDriver :)

Comment (ytimocin) [reply pkg/recipes/terraform/module.go:102]: This way we could also mock some of the functions of **tfexec.Terraform**.

**Review by ytimocin — APPROVED** [review]

File: pkg/recipes/terraform/execute.go (hunk 1 above)
Comment (kachawla) [reply pkg/recipes/terraform/execute.go:86]: We are still creating working directory inside root directory https://github.com/radius-project/radius/blob/2dde25c6cfac1c54a5fe024c6a71f14d8420da5a/pkg/recipes/terraform/types.go#L67, this PR is just moving creation of working directory closer to initialization of terraform-exec. Did you mean something else?

File: pkg/recipes/terraform/module.go (hunk 2 above)
Comment (kachawla) [reply pkg/recipes/terraform/module.go:102]: That makes sense to me. Some of the functions called from execute.go can be cleaned up as well downloadModule, inspectModule etc, but it's best to do that in smaller PRs so didn't include that here.

**Review by sk593 — APPROVED** [review]

**Pull request merged**